// @restatedev/restate-sdk/dist/cjs/src/utils/completable_promise.js
"use strict";
/*
 * Copyright (c) 2023-2024 - Restate Software, Inc., Restate GmbH
 *
 * This file is part of the Restate SDK for Node.js/TypeScript,
 * which is released under the MIT license.
 *
 * You can find a copy of the license in file LICENSE in the root
 * directory of this repository or package, or at
 * https://github.com/restatedev/sdk-typescript/blob/main/LICENSE
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.CompletablePromise = void 0;
// Like https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Promise/withResolvers
// (not yet available in node)
class CompletablePromise {
    success;
    failure;
    promise;
    constructor() {
        this.promise = new Promise((resolve, reject) => {
            this.success = resolve;
            this.failure = reject;
        });
    }
    resolve(value) {
        this.success(value);
    }
    reject(reason) {
        this.failure(reason);
    }
}
exports.CompletablePromise = CompletablePromise;
