// @restatedev/restate-sdk/dist/cjs/src/common_api.js
"use strict";
/*
 * Copyright (c) 2023-2024 - Restate Software, Inc., Restate GmbH
 *
 * This file is part of the Restate SDK for Node.js/TypeScript,
 * which is released under the MIT license.
 *
 * You can find a copy of the license in file LICENSE in the root
 * directory of this repository or package, or at
 * https://github.com/restatedev/sdk-typescript/blob/main/LICENSE
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.TimeoutError = exports.TerminalError = exports.RestateError = exports.rpc = exports.SendOpts = exports.Opts = exports.handlers = exports.workflow = exports.object = exports.service = exports.serde = exports.CombineablePromise = void 0;
var context_js_1 = require("./context.js");
Object.defineProperty(exports, "CombineablePromise", { enumerable: true, get: function () { return context_js_1.CombineablePromise; } });
var restate_sdk_core_1 = require("@restatedev/restate-sdk-core");
Object.defineProperty(exports, "serde", { enumerable: true, get: function () { return restate_sdk_core_1.serde; } });
var rpc_js_1 = require("./types/rpc.js");
Object.defineProperty(exports, "service", { enumerable: true, get: function () { return rpc_js_1.service; } });
Object.defineProperty(exports, "object", { enumerable: true, get: function () { return rpc_js_1.object; } });
Object.defineProperty(exports, "workflow", { enumerable: true, get: function () { return rpc_js_1.workflow; } });
Object.defineProperty(exports, "handlers", { enumerable: true, get: function () { return rpc_js_1.handlers; } });
Object.defineProperty(exports, "Opts", { enumerable: true, get: function () { return rpc_js_1.Opts; } });
Object.defineProperty(exports, "SendOpts", { enumerable: true, get: function () { return rpc_js_1.SendOpts; } });
var rpc_js_2 = require("./types/rpc.js");
Object.defineProperty(exports, "rpc", { enumerable: true, get: function () { return rpc_js_2.rpc; } });
var errors_js_1 = require("./types/errors.js");
Object.defineProperty(exports, "RestateError", { enumerable: true, get: function () { return errors_js_1.RestateError; } });
Object.defineProperty(exports, "TerminalError", { enumerable: true, get: function () { return errors_js_1.TerminalError; } });
Object.defineProperty(exports, "TimeoutError", { enumerable: true, get: function () { return errors_js_1.TimeoutError; } });
