// @restatedev/restate-sdk-core/dist/cjs/src/serde_api.js
"use strict";
/*
 * Copyright (c) 2023-2024 - Restate Software, Inc., Restate GmbH
 *
 * This file is part of the Restate SDK for Node.js/TypeScript,
 * which is released under the MIT license.
 *
 * You can find a copy of the license in file LICENSE in the root
 * directory of this repository or package, or at
 * https://github.com/restatedev/sdk-typescript/blob/main/LICENSE
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.serde = void 0;
class JsonSerde {
    contentType = "application/json";
    serialize(value) {
        if (value === undefined) {
            return new Uint8Array(0);
        }
        return new TextEncoder().encode(JSON.stringify(value));
    }
    deserialize(data) {
        if (data.length === 0) {
            return undefined;
        }
        return JSON.parse(new TextDecoder().decode(data));
    }
}
class BinarySerde {
    contentType = "application/octet-stream";
    serialize(value) {
        return value;
    }
    deserialize(data) {
        return data;
    }
}
class VoidSerde {
    contentType = "application/octet-stream";
    serialize(value) {
        if (value !== undefined) {
            throw new Error("Expected undefined value");
        }
        return new Uint8Array(0);
    }
    deserialize(data) {
        if (data.length !== 0) {
            throw new Error("Expected empty data");
        }
    }
}
var serde;
(function (serde) {
    serde.json = new JsonSerde();
    serde.binary = new BinarySerde();
    serde.empty = new VoidSerde();
})(serde || (exports.serde = serde = {}));
