// @restatedev/restate-sdk/dist/cjs/src/types/discovery.js
"use strict";
/*
 * Copyright (c) 2023-2024 - Restate Software, Inc., Restate GmbH
 *
 * This file is part of the Restate SDK for Node.js/TypeScript,
 * which is released under the MIT license.
 *
 * You can find a copy of the license in file LICENSE in the root
 * directory of this repository or package, or at
 * https://github.com/restatedev/sdk-typescript/blob/main/LICENSE
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.ServiceHandlerType = exports.ServiceType = exports.ProtocolMode = void 0;
var ProtocolMode;
(function (ProtocolMode) {
    ProtocolMode["BIDI_STREAM"] = "BIDI_STREAM";
    ProtocolMode["REQUEST_RESPONSE"] = "REQUEST_RESPONSE";
})(ProtocolMode || (exports.ProtocolMode = ProtocolMode = {}));
var ServiceType;
(function (ServiceType) {
    ServiceType["VIRTUAL_OBJECT"] = "VIRTUAL_OBJECT";
    ServiceType["SERVICE"] = "SERVICE";
    ServiceType["WORKFLOW"] = "WORKFLOW";
})(ServiceType || (exports.ServiceType = ServiceType = {}));
var ServiceHandlerType;
(function (ServiceHandlerType) {
    ServiceHandlerType["WORKFLOW"] = "WORKFLOW";
    ServiceHandlerType["EXCLUSIVE"] = "EXCLUSIVE";
    ServiceHandlerType["SHARED"] = "SHARED";
})(ServiceHandlerType || (exports.ServiceHandlerType = ServiceHandlerType = {}));
