// @restatedev/restate-sdk/dist/esm/src/types/discovery.js
/*
 * Copyright (c) 2023-2024 - Restate Software, Inc., Restate GmbH
 *
 * This file is part of the Restate SDK for Node.js/TypeScript,
 * which is released under the MIT license.
 *
 * You can find a copy of the license in file LICENSE in the root
 * directory of this repository or package, or at
 * https://github.com/restatedev/sdk-typescript/blob/main/LICENSE
 */
export var ProtocolMode;
(function (ProtocolMode) {
    ProtocolMode["BIDI_STREAM"] = "BIDI_STREAM";
    ProtocolMode["REQUEST_RESPONSE"] = "REQUEST_RESPONSE";
})(ProtocolMode || (ProtocolMode = {}));
export var ServiceType;
(function (ServiceType) {
    ServiceType["VIRTUAL_OBJECT"] = "VIRTUAL_OBJECT";
    ServiceType["SERVICE"] = "SERVICE";
    ServiceType["WORKFLOW"] = "WORKFLOW";
})(ServiceType || (ServiceType = {}));
export var ServiceHandlerType;
(function (ServiceHandlerType) {
    ServiceHandlerType["WORKFLOW"] = "WORKFLOW";
    ServiceHandlerType["EXCLUSIVE"] = "EXCLUSIVE";
    ServiceHandlerType["SHARED"] = "SHARED";
})(ServiceHandlerType || (ServiceHandlerType = {}));
