// @restatedev/restate-sdk/dist/esm/src/endpoint/lambda_endpoint.js
/*
 * Copyright (c) 2023-2024 - Restate Software, Inc., Restate GmbH
 *
 * This file is part of the Restate SDK for Node.js/TypeScript,
 * which is released under the MIT license.
 *
 * You can find a copy of the license in file LICENSE in the root
 * directory of this repository or package, or at
 * https://github.com/restatedev/sdk-typescript/blob/main/LICENSE
 */
import { EndpointBuilder } from "./endpoint_builder.js";
import { GenericHandler } from "./handlers/generic.js";
import { LambdaHandler } from "./handlers/lambda.js";
import { ProtocolMode } from "../types/discovery.js";
export class LambdaEndpointImpl {
    builder = new EndpointBuilder();
    get keySet() {
        return this.builder.keySet;
    }
    componentByName(componentName) {
        return this.builder.componentByName(componentName);
    }
    addComponent(component) {
        this.builder.addComponent(component);
    }
    bindBundle(services) {
        services.registerServices(this);
        return this;
    }
    bind(definition) {
        this.builder.bind(definition);
        return this;
    }
    withIdentityV1(...keys) {
        this.builder.withIdentityV1(...keys);
        return this;
    }
    setLogger(logger) {
        this.builder.setLogger(logger);
        return this;
    }
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    handler() {
        const genericHandler = new GenericHandler(this.builder, ProtocolMode.REQUEST_RESPONSE);
        const lambdaHandler = new LambdaHandler(genericHandler);
        return lambdaHandler.handleRequest.bind(lambdaHandler);
    }
}
