// @restatedev/restate-sdk/dist/esm/src/endpoint/node_endpoint.js
/*
 * Copyright (c) 2023-2024 - Restate Software, Inc., Restate GmbH
 *
 * This file is part of the Restate SDK for Node.js/TypeScript,
 * which is released under the MIT license.
 *
 * You can find a copy of the license in file LICENSE in the root
 * directory of this repository or package, or at
 * https://github.com/restatedev/sdk-typescript/blob/main/LICENSE
 */
import * as http2 from "http2";
import { LambdaHandler } from "./handlers/lambda.js";
import { EndpointBuilder } from "./endpoint_builder.js";
import { GenericHandler } from "./handlers/generic.js";
import { Readable, Writable } from "node:stream";
import { ProtocolMode } from "../types/discovery.js";
import { ensureError } from "../types/errors.js";
export class NodeEndpoint {
    builder = new EndpointBuilder();
    get keySet() {
        return this.builder.keySet;
    }
    componentByName(componentName) {
        return this.builder.componentByName(componentName);
    }
    addComponent(component) {
        this.builder.addComponent(component);
    }
    bindBundle(services) {
        services.registerServices(this);
        return this;
    }
    bind(definition) {
        this.builder.bind(definition);
        return this;
    }
    withIdentityV1(...keys) {
        this.builder.withIdentityV1(...keys);
        return this;
    }
    setLogger(logger) {
        this.builder.setLogger(logger);
        return this;
    }
    http2Handler() {
        const handler = new GenericHandler(this.builder, ProtocolMode.BIDI_STREAM);
        return (request, response) => {
            (async () => {
                try {
                    const url = request.url;
                    const resp = await handler.handle({
                        url,
                        headers: request.headers,
                        body: Readable.toWeb(request),
                        extraArgs: [],
                    });
                    response.writeHead(resp.statusCode, resp.headers);
                    const responseWeb = Writable.toWeb(response);
                    await resp.body.pipeTo(responseWeb);
                    await new Promise((resolve) => response.end(resolve));
                }
                catch (e) {
                    const error = ensureError(e);
                    this.builder.rlog.error("Error while handling connection: " + (error.stack ?? error.message));
                    response.destroy(error);
                }
            })().catch(() => { });
        };
    }
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    lambdaHandler() {
        const genericHandler = new GenericHandler(this.builder, ProtocolMode.REQUEST_RESPONSE);
        const handler = new LambdaHandler(genericHandler);
        return handler.handleRequest.bind(handler);
    }
    listen(port) {
        const actualPort = port ?? parseInt(process.env.PORT ?? "9080");
        this.builder.rlog.info(`Listening on ${actualPort}...`);
        const server = http2.createServer(this.http2Handler());
        return new Promise((resolve, reject) => {
            let failed = false;
            server.once("error", (e) => {
                failed = true;
                reject(e);
            });
            server.listen(actualPort, () => {
                if (failed) {
                    return;
                }
                const address = server.address();
                if (address === null || typeof address === "string") {
                    reject(new TypeError("endpoint.listen() currently supports only binding to a PORT"));
                }
                else {
                    resolve(address.port);
                }
            });
        });
    }
}
